Terminal.Gui/Core/View.cs

```diff
@@ -1354,26 +1354,28 @@ public class FocusEventArgs : EventArgs {
 		/// <summary>
 		/// Method invoked when a subview is being added to this view.
 		/// </summary>
-		/// <param name="view">The subview being added.</param>
-		public virtual void OnAdded (View view)
+		/// <param name="e">Event where <see cref="ViewEventArgs.View"/> is the subview being added.</param>
+		public virtual void OnAdded (ViewEventArgs e)
 		{
+			var view = e.View;
 			view.IsAdded = true;
 			view.x ??= view.frame.X;
 			view.y ??= view.frame.Y;
 			view.width ??= view.frame.Width;
 			view.height ??= view.frame.Height;
 
-			view.Added?.Invoke (this);
+			view.Added?.Invoke (this,e);
```

Comment:
Hmn so in the old version of `virtual void OnAdded (View view)` the `view` passed into this method was not passed on to `Invoke`.  Instead `this` was passed to `Invoke`.  

That means that the `view` reported by the `Added` event is the parent and not the child being added.

This is why the test is failing.  `Added_Removed` asserts the following:

```csharp
v.Added += (s,e) => {
	Assert.True (v.SuperView == e.View);
};
```

I propose we change this test to instead say

```csharp
v.Added += (s,e) => {
	Assert.True (v.SuperView == s);
	Assert.True (v == e.View);
};
```

Reply:
Actually I am confused about this test.  This is what it looks like in `v2_develop`

The description for the `Added` event says:

> Event fired when a subview is being added to this view.

But the test is registering the event on the child (`v`) and then calling add on the parent (`t`).

Is this to say then that `Added` event covers both

- When a subview is added to yourself
- When you are added as a subview to someone else?

```csharp
public void Added_Removed ()
{
	var v = new View (new Rect (0, 0, 10, 24));
	var t = new View ();

	v.Added += (View e) => {
		Assert.True (v.SuperView == e);
	};

	v.Removed += (View e) => {
		Assert.True (v.SuperView == null);
	};

	t.Add (v);
	Assert.True (t.Subviews.Count == 1);

	t.Remove (v);
	Assert.True (t.Subviews.Count == 0);
}
```

Reply:
I agree. This event is normally used by the subview to get notified when the superview has add it and so your logic is perfectly right. By the way thanks for great effort.

Reply:
> * When a subview is added to yourself
> 
> * When you are added as a subview to someone else?

Ok this is not the case.  I think the description and my expectation is in fact wrong.  Based on the `Add` method `Added` event is fired on a `View` **when it is aded to a parent**.  And **not** "when a subview is being added" [to it].

```
public virtual void Add (View view)
{
	if (view == null)
		return;
        [...]
	OnAdded (view);
```

Reply:
Yes, that why it's invoke `view.Added?.Invoke (this);`

Reply:
I propose this change, because it the user use the arg passed will think it's the  superview and that why the test is failing.

```cs
public virtual void OnAdded (ViewEventArgs e)
{
	var view = e.View;
	view.IsAdded = true;
	view.x ??= view.frame.X;
	view.y ??= view.frame.Y;
	view.width ??= view.frame.Width;
	view.height ??= view.frame.Height;
	e.View = this;
	view.Added?.Invoke (this, e);
}
```

Reply:
The same for the `OnRemoved`;

Reply:
Ok I  have added a new `EventArgs`  called `SuperViewChangedEventArgs` so it is clear to user what view is being talked about (since it contains both `Parent` and `Child`).  Test now passes as:

```csharp
[Fact]
public void Added_Removed ()
{
	var v = new View (new Rect (0, 0, 10, 24));
	var t = new View ();

	v.Added += (s,e) => {
		Assert.Same (v.SuperView, e.Parent);
		Assert.Same (t, e.Parent);
		Assert.Same (v, e.Child);
	};

	v.Removed += (s, e) => {
		Assert.Same (t, e.Parent);
		Assert.Same (v, e.Child);
		Assert.True (v.SuperView == null);
	};

	t.Add (v);
	Assert.True (t.Subviews.Count == 1);

	t.Remove (v);
	Assert.True (t.Subviews.Count == 0);
}
```

Reply:
Much better solution, congrats for your brilliant implementation.